[KOGITO-256] -Kogito CLI to upload source files from local machine to… (#247)
kiegroup/kogito-cloud-operator: README.md

feature (available only on OpenShift). This kind of build consists id upload a Kogito service file, e.g. a dmn or a bpmn file
to the OpenShift Cluster and trigger a new Source to Image build automatically.

You can provide a single dmn, drl or bpmn|bpmn2 files, the standalone file can be locally on the filesystem or on a Web Site, like Git Hub:

```bash
# from filesystem
$ kogito -service example-dmn-quarkus --from-file TrafficViolation.dmn
File found /tmp/kogito-test/src/main/resources/TrafficViolation.dmn.
...
The requested file(s) was successfully upload to OpenShift, a build with this file should now be running. To see its logs, run 'oc logs -f bc/test-builder -n kogito'
```

While checking OpenShift build logs with this command **oc logs -f bc/test-builder -n kogito** you can see on the beginning ot the build log:
```
Receiving source from STDIN as file TrafficViolation.dmn
Using docker-registry.default.svc:5000/spolti-1/kogito-quarkus-ubi8-s2i@sha256:729e158710dedba50a49943ba188d8f31d09568634896de9b903838fc4e34e94 as the s2i builder image
```

```bash
# from a git repository

> $ kogito -service example-dmn-quarkus --from-file https://raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic%20Violation.dmn

Author: LOL, again hahaha
dam ctrl c and v

...
The requested file(s) was successfully upload to OpenShift, a build with this file should now be running. To see its logs, run 'oc logs -f bc/test-builder -n kogito'
```

If you check the logs, we can also see the dmn file being used to to build the Kogito service.

Sometimes we also need to provide more than one Kogito resource file, e.g. a drl and a bpmn. In that case all we need to do

> is, instead specify a directory instead, if any valid file is found on the given directory, the cli will compress these

Reviewer: > if any valid file is found on the given directory

So, only files with specific suffix are copied?

Author: yes, IMHO there's no point to allow user to push any kind of file since the archetypes can not work properly.

Author: does it make sense?

Author: it can be improved in the future if needed.
README updated.

> files and upload to the OpenShift cluster. To start upload various files, put all the desired files in a directory and 

Author: is 'To upload more than one file' better?

> specify it as parameter. Note that if you have more files on this directory, they wont be copied.

Reviewer: > Note that if you have more files on this directory, they wont be copied.

What does it mean? From previous sentences I understood that more files can be uploaded at once.

Author: it is not clear, updated.

```bash

> kogito --verbose deploy-service test --from-file /home/user/development/kogito-test/

Author: no need, probably leftover.

```

> If for some reason the build fails, pass the env BUILD_LOGLEVEL with the desired level, i.e. 9 as a build-env parameter, i.e.:

Author: +1

```bash

> kogito --verbose deploy-service test --from-file /home/user/development/kogito-test/  --build-env BUILD_LOGLEVEL=9

Author: here it is intentional, the cli verbose can bring useful info.

And that's it. In a couple minutes you should have your Kogito Service application up and running using the same binaries
built locally.

### Build From File

Kogito Operator can configure a build for you that will pull and build your application through the source to image (s2i)
feature (available only on OpenShift). This kind of build consists to upload a Kogito service file, e.g. a dmn or a bpmn file
to the OpenShift Cluster and trigger a new Source to Image build automatically.

You can provide a single dmn, drl, bpmn or bpmn2 files, the standalone file can be locally on the filesystem or on a Web Site, like Git Hub:

> 

Reviewer 2: There is an extra empty line here.

```bash
# from filesystem

> $ kogito deploy-service example-dmn-quarkus --from-file TrafficViolation.dmn

Reviewer 2: I think it would be good to use the same source for filesystem and from git repository. For example, the TrafficViolation.dmn file does not exist, but "Traffic Violation.dmn".

Author: Maybe for windows users?
At least on Unix systems it is not common to use name files with spaces.

We can change it.

@radtriste @sutaakar what do you guys think?

Reviewer 3: > I think it would be good to use the same source for filesystem and from git repository

I didn't understand this question. Doesn't matter if the file has space or not, the plugin will create the project anyway. What we can do is run a regexp in the file name and emit a warn or deny the upload based on this rule.

Reviewer 3: Oh you guys are talking about the README description of the example. Right, agreed.

Author: using spaces from now.

> $ kogito deploy-service example-dmn-quarkus --from-file TrafficViolation.dmn

Reviewer 4: ```suggestion
$ kogito deploy-service example-dmn-quarkus --from-file "Traffic Violation.dmn"
```

> File found /tmp/kogito-test/src/main/resources/TrafficViolation.dmn.

Reviewer 4: ```suggestion
File found /tmp/kogito-examples/dmn-quarkus-example/src/main/resources/Traffic%20Violation.dmn.
```

...
The requested file(s) was successfully upload to OpenShift, a build with this file should now be running. To see its logs, run 'oc logs -f bc/example-dmn-quarkus -n kogito'
```

While checking OpenShift build logs with this command **oc logs -f bc/test-builder -n kogito** you can see on the beginning ot the build log:
```
Receiving source from STDIN as file TrafficViolation.dmn
Using docker-registry.default.svc:5000/spolti-1/kogito-quarkus-ubi8-s2i@sha256:729e158710dedba50a49943ba188d8f31d09568634896de9b903838fc4e34e94 as the s2i builder image
```

```bash
# from a git repository
$ kogito deploy-service example-dmn-quarkus --from-file https://raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic%20Violation.dmn
...

> The requested file(s) was successfully upload to OpenShift, a build with this file should now be running. To see its logs, run 'oc logs -f bc/test-builder -n kogito'

Reviewer 4: ```suggestion
The requested file(s) was successfully upload to OpenShift, a build with this file should now be running. To see its logs, run 'oc logs -f bc/example-dmn-quarkus-builder -n kogito'
```

```

> If you check the logs, we can also see the dmn file being used to to build the Kogito service.

Reviewer 4: ```suggestion
If you check the logs, we can also see the dmn file being used to build the Kogito service.
```

> Sometimes we also need to provide more than one Kogito resource file, e.g. a drl and a bpmn. IIn that case, 

Reviewer 2: Typo here: "IIn"

> Sometimes we also need to provide more than one Kogito resource file, e.g. a drl and a bpmn. IIn that case, 

Reviewer 4: ```suggestion
Sometimes we also need to provide more than one Kogito resource file, e.g. a drl and a bpmn. In that case, 
```

all we need to do is to specify a whole directory. If any valid file is found in the given directory, the cli will compress these

> files and upload to the OpenShift cluster. To upload more than one file, put all the desired files in a directory and 

Reviewer 2: This line must be appended to the previous line.

> specify it as parameter. Note that if you have other unsupported files in this directory, they will not be copied.

Reviewer 4: ```suggestion
specify it as parameter.  
Note that if you have other unsupported files in this directory, they will not be copied.
```

```bash

> kogito deploy-service test --from-file /home/user/development/kogito-test/

Reviewer 2: I have a question about uploading folders. Do we support also Java sources like models? As far I can see below, we do not.

Author: no, only bpmn, drl or dmn.

```

As output of the cli command above, we will have something similar to:

```bash
The provided source is a dir, packing files.
...

> The requested file(s) was successfully upload to OpenShift, a build with this file should now be running. To see its logs, run 'oc logs -f bc/test-builder -n kogito'

Reviewer 4: ```suggestion
The requested file(s) was successfully upload to OpenShift, a build with the file(s) should now be running. To see its logs, run 'oc logs -f bc/test-builder -n kogito'
```

```

> If for some reason the build fails, pass the env BUILD_LOGLEVEL with the desired level, e.g. 9 as a build-env parameter:

Reviewer 2: Can we make the "BUILD_LOGLEVEL" a link? Maybe: https://docs.openshift.com/container-platform/3.5/dev_guide/builds/basic_build_operations.html. As far I could see, the level 9 does not exist, the range is between 1 to 5.

Author: surething :)

```bash
kogito --verbose deploy-service test --from-file /home/user/development/kogito-test/  --build-env BUILD_LOGLEVEL=9
```

> 

Reviewer 2: Extra blank line here.

And that's it. In a couple minutes you should have your Kogito Service application up and running using the same binaries
built locally.

### Build From File

Kogito Operator can configure a build for you that will pull and build your application through the source to image (s2i)
feature (available only on OpenShift). This kind of build consists to upload a Kogito service file, e.g. a dmn or a bpmn file
to the OpenShift Cluster and trigger a new Source to Image build automatically.

You can provide a single dmn, drl, bpmn, bpmn2 and properties files, the standalone file can be locally on the filesystem or on a Web Site, like GitHub:

```bash
# from filesystem
$ kogito deploy-service example-dmn-quarkus --from-file  /tmp/kogito-examples/dmn-quarkus-example/src/main/resources/"Traffic Violation.dmn"
File found /tmp/kogito-examples/dmn-quarkus-example/src/main/resources/Traffic Violation.dmn.
...

> The requested file(s) was successfully upload to OpenShift, a build with this file(s) should now be running. To see the logs, run 'oc logs -f bc/example-dmn-quarkus-builder -n kogito'

Reviewer 5: "upload**ed** to openshift". Should fix in docs, and source code.

Author: done.
